### PR TITLE
AIRFLOW-124 Implement create_dagrun

### PR DESCRIPTION
**Pull request by bolkedebruin**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- AIRFLOW-124

This PR forms the basis for the roadmap (https://drive.google.com/open?id=0B_Y7S4YFVWvYM1o0aDhKMjJhNzg) . It is one of the initial commits from https://github.com/apache/incubator-airflow/compare/master...bolkedebruin:AIRFLOW_SCHEDULER .

**DAG.create_dagrun (please review: @aoen, @jlowin, @mistercrunch, @artwr )**

This creates dagrun from a Dag. It also creates the TaskInstances from the tasks known at instantiation time. By having taskinstances created at dagrun instantiation time, deadlocks that were tested for will not take place anymore (@jlowin, correct? different test required?). For now, the visual consequence of having these taskinstances already there is that they will be black in the tree view.

Tests in core.py were adjusted as they were supposedly creating a dagrun with tasks, while they were actually creating dagruns and orphaned TaskInstances (ie. the dag_id was not matching the dag_id from the dagrun). This was discussed with @artwr, who said these were remnants from the past.

By doing this I also fixed an issue in models.py that a dag was not set for a task if called from dag.add_task (@aoen, @jlowin).

**DagRun.find is a convenience function that returns the DagRuns for a given dag. It makes sure to have a single place how to find dagruns**

DagRun.find does not work for all cases yet (ie multiple execution dates etc). My aim is to limit the ORM required in sub functions that implement somethings just a bit differently creating compatibility issues.

### CODE REVIEW

File: airflow/models.py [hunk 1]
```diff
@@ -3368,6 +3416,37 @@ def __repr__(self):
...
 
+    @provide_session
+    def refresh_from_db(self, session=None):
+        DR = DagRun
+
+        dr = session.query(DR).filter(
+            DR.dag_id == self.dag_id,
+            DR.execution_date == self.execution_date,
+            DR.run_id == self.run_id
+        ).first()
```
Comment (criccomini) [inline airflow/models.py]: Should there only ever by one dag run with a given dag_id/execution_date/run_id? If so, wonder if we want to fail, or log grumpily if we get more than one response back.

File: airflow/models.py [hunk 2]
```diff
@@ -3368,6 +3416,37 @@ def __repr__(self):
     def id_for_date(klass, date, prefix=ID_FORMAT_PREFIX):
         return prefix.format(date.isoformat()[:19])
 
+    @provide_session
+    def refresh_from_db(self, session=None):
```
Comment (criccomini) [inline airflow/models.py:3423]: Nit: docs

File: airflow/models.py [hunk 3]
```diff
@@ -3368,6 +3416,37 @@ def __repr__(self):
...
+            DR.dag_id == self.dag_id,
+            DR.execution_date == self.execution_date,
+            DR.run_id == self.run_id
+        ).first()
+        if dr:
+            self.id = dr.id
+            self.state = dr.state
+
+    @staticmethod
+    def find(dag_id, run_id=None, state=None, external_trigger=None):
```
Comment (criccomini) [inline airflow/models.py]: Nit: docs

**Comment by criccomini** [comment]
Couple of nits/questions, but LGTM overall. Take my opinion with a grain of salt, since my scheduler knowledge is unfortunately a bit lacking.

File: airflow/models.py [hunk 4]
```diff
@@ -3073,6 +3074,53 @@ def cli(self):
...
+                      conf=None,
+                      session=None):
+        """
+        Creates a dag run from this dag including the tasks associated with this dag. Returns the dag
+        run.
+        :param run_id: defines the the run id for this dag run
+        :type run_id: string
+        :param execution_date: the execution date of this dag run
+        :type execution_date: datetime
+        :param start_date: the date this dag run should be evaluated
```
Comment (aoen) [inline airflow/models.py]: Nit: order of params is backwards (state should come before start_date), and conf/session are missing

File: airflow/bin/cli.py [hunk 5]
```diff
@@ -118,29 +118,32 @@ def backfill(args, dag=None):
 
 
 def trigger_dag(args):
-    session = settings.Session()
-    # TODO: verify dag_id
+    dag = get_dag(args)
+
+    if not dag:
+        logging.error("Cannot find dag {}".format(args.dag_id))
+        sys.exit(1)
```
Comment (aoen) [inline airflow/bin/cli.py:125]: I wonder if we should use different exit codes for different errors statuses, also it looks like the other parts of the CLI raise instead of exiting (though exiting is the right call since this is user facing), maybe turn the raises in the top-level functions into sys.exits too to be consistent?

File: airflow/models.py [hunk 6]
```diff
@@ -3073,6 +3074,53 @@ def cli(self):
...
+            execution_date=execution_date,
+            start_date=start_date,
+            external_trigger=external_trigger,
+            conf=conf,
+            state=state
+        )
+        session.add(run)
+
+        for task in self.tasks:
+            # todo: check this
```
Comment (aoen) [inline airflow/models.py]: Nit: Consider elaborating on what this means

File: airflow/models.py [hunk 7]
```diff
@@ -3368,6 +3416,37 @@ def __repr__(self):
...
+        ).first()
+        if dr:
+            self.id = dr.id
+            self.state = dr.state
+
+    @staticmethod
+    def find(dag_id, run_id=None, state=None, external_trigger=None):
+        DR = DagRun
+
+        session = settings.Session()
```
Comment (aoen) [inline airflow/models.py]: Why not use the @provide_session decorator here instead?

File: airflow/models.py (hunk 1 above)
Comment (bolkedebruin) [inline airflow/models.py]: It refreshes itself so no we cannot have more. Maybe one() would be better.

File: airflow/bin/cli.py (hunk 5 above)
Comment (bolkedebruin) [inline airflow/bin/cli.py:125]: @aoen I agree, but the full change should be part of a different PR I think. I can change to raising and create a Jira for it?

File: airflow/models.py [hunk 8]
```diff
@@ -3073,6 +3074,53 @@ def cli(self):
...
+            # todo: check this
+            if task.adhoc:
+                continue
+
+            ti = TaskInstance(task, execution_date)
+            session.add(ti)
+
+        session.commit()
+
+        run.refresh_from_db()
```
Comment (aoen) [inline airflow/models.py:3124]: What's the purpose of this (and maybe consider adding a comment in the code too with your answer)?

File: airflow/models.py (hunk 6 above)
Comment (bolkedebruin) [inline airflow/models.py]: I wasn't sure myself, but the recent discussion on the mailinglist is helping out here :)

File: airflow/bin/cli.py (hunk 5 above)
Comment (aoen) [inline airflow/bin/cli.py:125]: SGTM, and thanks!

File: airflow/models.py (hunk 7 above)
Comment (bolkedebruin) [inline airflow/models.py]: @aoen I don't like the decorator very much. A typical ORM session should look like

session.open()
try:
   session.commit
catch:
   session.rollback
finally:
   session.close

The decorator does not encourage this behavior: it makes it hard to deal with errors. But here I could use it.

File: airflow/models.py (hunk 6 above)
Comment (aoen) [inline airflow/models.py]: Should have been more clear sorry, I meant what the TODO is for, not the adhoc tasks (unless this is a comment you put in as a reminder to yourself to fill this in before merging the PR in which case ignore me).

File: airflow/models.py (hunk 7 above)
Comment (aoen) [inline airflow/models.py]: Fair point. I think we should be consistent throughout and go with one way or another but that is out of scope.

File: airflow/bin/cli.py (hunk 5 above)
Comment (bolkedebruin) [inline airflow/bin/cli.py:125]: https://issues.apache.org/jira/browse/AIRFLOW-126

File: airflow/models.py (hunk 6 above)
Comment (bolkedebruin) [inline airflow/models.py]: It was a todo to check why I needed to check for adhoc tasks (this came verbatim from schedule_dag). So it was entirely clear what you were asking ;-)

File: airflow/models.py (hunk 8 above)
Comment (bolkedebruin) [inline airflow/models.py:3124]: Ah the important bit: this create_dagrun creates the _full_ dagrun and thus includes creating the taskinstances for this dagrun. Considering that a Dag is an ordered combination of tasks a DagRun should also be this at any given point in time. Ie. the current situation violates this: DagRuns are created shallow, without TaskInstances.

Having this in resolves issues with some occurrences of deadlocks and depend_on_past and I do consider it important for the integrity of airflow going forward and is one of the reasons of using "create_dagrun" to make sure we will be consistent in creating dag runs.

state for the create TaskInstances will be None / NULL and that does not interfere with any of the checks (confirmed by inspection and tests)

And yes I will add a comment in the code.

**Comment by jlowin** [comment]
@bolkedebruin LGTM

File: airflow/models.py (hunk 8 above)
Comment (aoen) [inline airflow/models.py:3124]: Makes sense, thanks for the thorough explanation!

**Comment by aoen** [comment]
LGTM after commented added

File: airflow/models.py [hunk 9]
```diff
@@ -3368,6 +3419,50 @@ def __repr__(self):
...
+            DR.execution_date == self.execution_date,
+            DR.run_id == self.run_id
+        ).one()
+        if dr:
+            self.id = dr.id
+            self.state = dr.state
+
+    @staticmethod
+    @provide_session
+    def find(dag_id, run_id=None, state=None, external_trigger=None, session=None):
```
Comment (artwr) [inline airflow/models.py:3441]: Thanks for encapsulating the logic by the way, this is amazing.

File: airflow/models.py (hunk 2 above)
Comment (mistercrunch) [inline airflow/models.py:3423]: about docstrings, I had pylint setup on landscape.io to only require a docstring if the method is 10+ lines which I thought was reasonable

File: airflow/models.py [hunk 10]
```diff
@@ -3368,6 +3419,50 @@ def __repr__(self):
...
+        Reloads the current dagrun from the database
+        :param session: database session
+        """
+        DR = DagRun
+
+        dr = session.query(DR).filter(
+            DR.dag_id == self.dag_id,
+            DR.execution_date == self.execution_date,
+            DR.run_id == self.run_id
+        ).one()
```
Comment (mistercrunch) [inline airflow/models.py:3434]: oh first time I see `one()`. sqla has a huge api... I just looked it up and `one()` raises if empty, meaning the if bellow is not necessary.

**Comment by mistercrunch** [comment]
LGTM except for the unnecessary `if` under the sqlaclehmy's `one()` call
http://docs.sqlalchemy.org/en/rel_1_0/orm/query.html#sqlalchemy.orm.query.Query.one

File: airflow/models.py (hunk 10 above)
Comment (bolkedebruin) [inline airflow/models.py:3434]: Ah yes, I was wondering why it wasn't used a bit more. Kept the 'if' in for consistency, but obviously it should be removed. I will provide a small update for that

**Comment by bolkedebruin** [comment]
@mistercrunch thanks. There is one operational regression (also heads up @aoen, @jlowin, @artwr) due to the eager creation of task instances. In its current form the scheduler will evaluate all State.NONE task instances and thus due to eager creation it will take more time to evaluate them. 

I'm addressing this issue in a follow up PR (almost done/ready for review).
